working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-containing a variable matcher with the same name as the field. The variable is
-always `final`. These cases are equivalent:
+When you want an `if` statement in an expression context, you can use a
+conditional expression (`?:`). There is no expression form for multi-way
+branching, so we define a new switch expression. It takes code like this:
 
 ```dart
-switch (obj) {
-  case (first: final first, second: final second): ...
-  case (first:, second:): ...
+Color shiftHue(Color color) {
+  switch (color) {
+    case Color.red:
+      return Color.orange;
+    case Color.orange:
+      return Color.yellow;
+    case Color.yellow:
+      return Color.green;
+    case Color.green:
+      return Color.blue;
+    case Color.blue:
+      return Color.purple;
+    case Color.purple:
+      return Color.red;
+  }
 }
 ```
 
-**TODO: Add a `...` syntax to allow ignoring positional fields?**
+And turns it into:
 
-#### Variable matcher
+```dart
+Color shiftHue(Color color) {
+  return switch (color) {
+    case Color.red => Color.orange;
+    case Color.orange => Color.yellow;
+    case Color.yellow => Color.green;
+    case Color.green => Color.blue;
+    case Color.blue => Color.purple;
+    case Color.purple => Color.red;
+  };
+}
+```
 
-A variable matcher lets a matching pattern also perform variable binding.
+The grammar is:
 
 ```
-variableMatcher ::= ( 'final' | 'var' | 'final'? typeWithBinder ) identifier
-```
+primary               ::= // Existing productions...
+                        | switchExpression
 
-By using variable matchers as subpatterns of a larger matched pattern, a single
-composite pattern can validate some condition and then bind one or more
-variables only when that condition holds.
-
-A variable pattern can also have a type annotation in order to only match values
-of the specified type.
+switchExpression      ::= 'switch' '(' expression ')' '{'
+                          switchExpressionCase* defaultExpressionCase? '}'
+switchExpressionCase  ::= caseHead '=>' expression ';'
+defaultExpressionCase ::= 'default' '=>' expression ';'
+```
 
-#### Extractor matcher
+Slotting into `primary` means it can be used anywhere any expression can appear,
+even as operands to unary and binary operators. Many of these uses are ugly, but
+not any more problematic than using a collection literal in the same context
+since a `switch` expression is always delimited by a `switch` and `}`.
 
-An extractor combines a type test and record destructuring. It matches if the
-object has the named type. If so, it then uses the following record pattern to
-destructure fields on the value as that type. This pattern is particularly
-useful for writing code in an algebraic datatype style. For example:
+Making it high precedence allows useful patterns like:
 
 ```dart
-class Rect {
-  final double width, height;
-
-  Rect(this.width, this.height);
-}
+await switch (n) {
+  case 1 => aFuture;
+  case 2 => anotherFuture;
+};
```

Comment:
These expression switches are invalid because they are not exhaustive.

Reply:
Fixed.